crates/ra_arena/src/map.rs

```diff
@@ -14,14 +14,9 @@ pub struct ArenaMap<ID, V> {
 impl<T, V> ArenaMap<Idx<T>, V> {
     pub fn insert(&mut self, id: Idx<T>, t: V) {
         let idx = Self::to_idx(id);
-        if self.v.capacity() <= idx {
-            self.v.reserve(idx + 1 - self.v.capacity());
-        }
-        if self.v.len() <= idx {
-            while self.v.len() <= idx {
-                self.v.push(None);
-            }
-        }
+
+        let fill = (idx + 1).saturating_sub(self.v.len());
+        self.v.extend(std::iter::repeat_with(|| None).take(fill));
```

Comment:
```suggestion
        self.v.resize(idx + 1, None);
```

Reply:
Sorry for the imperative rain on your functional parade :)

Reply:
ah, wait, it indeed can truncate, I guess that needs to be `(idx + 1).max(self.v.len())`